### PR TITLE
service/dap: fix closing closed channel bug

### PR DESCRIPTION
**Pull request by suzmue**
Our previous implementation of waiting on resumeRequestLoop was not
safe, because the channel may be closed several times. This change
removes this bug by instead having two channels to track the two
conditions that we are managing, whether the command has started running
and whether the command has completed (on error or other).

Fixes go-delve/delve#2757

### CODE REVIEW

**Review by aarzilli — APPROVED** [review]

**Review by polinasok — CHANGES_REQUESTED** [review]
I believe we have the following cases when calling async handlers:
1. Debugger command is never called
- there was an error in dap layer set up
- the dap layer determined that it was not necessary to call the command (e.g. stopOnEntry)
- Before: command caller closed setup-done because debugger was never called
2. Debugger command is called, but fails
- Failed during setup
- Failed after successful set-up during execution
- Before: command caller closed setup-done when debugger forgot to (some corner case that might have gotten fixed now) or detected it as closed after it was closed by the debugger. 
3. Debugger command is called and succeeds
- Before: Debugger closed the channel and caller detected it as closed.

All closures occurred on the same goroutine, so select worked.

So what changed? The auto-resume logic added signal redirection with an additional goroutine, so now we could have the new goroutine close the channel when Command signaled setup completion and the per-request goroutine close the same channel at the same time when Command quickly returned? Do I have that right?

So this change separates the two types of done - done setting up inside of the command and done calling (or attempting to call) the command?

**Comment by suzmue** [comment]
> So what changed? The auto-resume logic added signal redirection with an additional goroutine, so now we could have the new goroutine close the channel when Command signaled setup completion and the per-request goroutine close the same channel at the same time when Command quickly returned? Do I have that right?

Yes this does happen because of the auto resume logic. Since we are adding a separate goroutine to wait for resumeNotify, when we actually call s.debugger.Command, the original goroutine and that other goroutine could try to close the channel at the same time.

**Comment by polinasok** [comment]
> > So what changed? The auto-resume logic added signal redirection with an additional goroutine, so now we could have the new goroutine close the channel when Command signaled setup completion and the per-request goroutine close the same channel at the same time when Command quickly returned? Do I have that right?
> 
> Yes this does happen because of the auto resume logic. Since we are adding a separate goroutine to wait for resumeNotify, when we actually call s.debugger.Command, the original goroutine and that other goroutine could try to close the channel at the same time.

The channel that is closed on two different goroutines is allowNextStateChange. At one point we had a discussion that with the introduction of changeStateMu we might not need to block the request loop on allowNextStateChange at all. Have you considered that approach instead of adding another channel?

**Comment by derekparker** [comment]
Just curious on the status of this @polinasok @suzmue.

**Comment by derekparker** [comment]
Should this PR still be considered for the next release?

**Comment by suzmue** [comment]
@derekparker I haven't looked at this in a bit, so I will go ahead and rebase it.

I'm not sure it is a high priority that we need to get it in asap. I don't think we have seen any reports of users actually encountering this bug, so I think it is ok to push it to the next release so @polinasok can come back and review

**Comment by derekparker** [comment]
@suzmue this still needs a rebase. Would we still like to try targeting this for 1.8.4?

**Comment by suzmue** [comment]
@derekparker Thanks for bumping this. I rebased the change, but need to take a little bit of time to get the context back on this change to address the remaining comments. Will do this by the end of the week.

File: service/dap/server.go [hunk 1]
```diff
@@ -3359,17 +3378,24 @@ func (s *Session) checkHaltRequested() bool {
...
 // of the target when the program is halted.
-func (s *Session) resumeOnce(command string, allowNextStateChange chan struct{}) (bool, *api.DebuggerState, error) {
+func (s *Session) resumeOnce(command string, allowNextStateChange *asyncSetupDone) (bool, *api.DebuggerState, error) {
 	// No other goroutines should be able to try to resume
 	// or halt execution while this goroutine is resuming
 	// execution, so we do not miss those events.
-	asyncSetupDone := make(chan struct{}, 1)
-	defer closeIfOpen(asyncSetupDone)
+	asyncSetupDone := newAsyncSetupDone()
+	defer close(asyncSetupDone.doneCh)
```
[the remark this answers is not in the document]
Comment (suzmue) [reply service/dap/server.go]: Done.

File: service/dap/server.go [hunk 2]
```diff
@@ -3359,17 +3378,24 @@ func (s *Session) checkHaltRequested() bool {
...
 	s.changeStateMu.Lock()
 	go func() {
-		defer s.changeStateMu.Unlock()
-		defer closeIfOpen(allowNextStateChange)
-		<-asyncSetupDone
+		// When we are done waiting for the command to finish or this function to complete,
+		// we want to close the allowNextStateChange.commandCh to notify the request loop
+		// that the setup is done and unlock s.changeStateMu.
```
[the remark this answers is not in the document]
Comment (suzmue) [reply service/dap/server.go]: Done.

File: service/dap/server.go [hunk 3]
```diff
@@ -3359,17 +3378,24 @@ func (s *Session) checkHaltRequested() bool {
...
+		defer func() {
+			if allowNextStateChange != nil {
```
[the remark this answers is not in the document]
Comment (suzmue) [reply service/dap/server.go:3457]: Done.

Yeah this isn't the most readable, even with this changes. hopefully the changes for handling logpoints could make this logic less complicated.

**Comment by polinasok** [comment]
We just hit this [again](https://delve.teamcity.com/buildConfiguration/Delve_AggregatorBuild/37670?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandPull+Request+Details=true&expandBuildTestsSection=true). I will take a look shortly.

**Comment by derekparker** [comment]
@suzmue @polinasok is this change still active or should I close it out?

**Pull request closed without merging**